### PR TITLE
week_6: Module C (The Librarian) — C.4 decision engine + golden-set decision gate

### PR DESCRIPTION
**Pull request by PRAteek-singHWY**
Hi @northdpole - Week 6 of Module C. Week 5 produced one **honest** confidence per chunk; this PR turns that number into the actual **decision** - auto-link into the graph, or route to a human - which is the accuracy gate of the whole pipeline.

> **Stacked on #974 (Week 5).** Based on `gsocmodule_C_week_5`; only the top commit is new. Until #974 merges the diff shows the W5 commits too - I'll rebase onto `main` as it lands, shrinking it to the **Week-6-only** surface (3 files). No dependency on Modules A or B: this runs entirely on the golden dataset.

## Overview

Week 3 built the **search step (C.1)**, Week 4 the **rerank step (C.2)**, and Week 5 the **calibration step (C.3)** - a single scalar `T` that maps the reranked shortlist to a trustworthy `confidence = softmax(logits / T)`.

**The problem:** a calibrated confidence is only useful if something *acts* on it. Auto-linking a wrong CRE pollutes the graph; sending everything to a human defeats the point. We need a rule that auto-links when it's safe and escalates when it isn't.

**This PR's role:** build the **decision step (C.4)** - `decision_engine.decide()`. It links the top-1 candidate iff `confidence >= threshold` **and** there is a candidate **and** no blocking safety flag; otherwise it routes to review with a `reason_code`. It's a pure function of `(confidence, candidates, flags, threshold) -> DecisionResult` - it does **not** import the C.3 scaler (confidence-in → decision-out), so it stays model-free and hermetically testable, mirroring the C.1/C.2/C.3 seams. Reason-code precedence is total: `NO_CANDIDATES > ADVERSARIAL_FLAG > UPDATE_AMBIGUOUS > BELOW_THRESHOLD`.

The harness gains a **decision-accuracy gate**: run the live C.1→C.4 decision over the golden set and measure how often `decide()` lands on the expected auto-link-vs-review call.

**Scope:** 1 new module + 1 new test + additive harness wiring. The SafetyGuard flags (`adversarial` / `update_ambiguous`) are accepted by `decide()` but not yet produced - nothing sets them, so they default to False (declared-degraded until that lands). No frontend, no migration, no behaviour change to OpenCRE proper.

## What changed

| Area | Files | Description |
|---|---|---|
| C.4 decision engine | `decision_engine.py` *(new)* | `decide(confidence, candidate_cre_ids, *, threshold, adversarial, update_ambiguous) -> DecisionResult`. Links the top-1 iff `confidence >= threshold` AND candidates exist AND no blocking flag; else reviews with a `reason_code`. Frozen `DecisionResult`, versioned `ENGINE_NAME`, custom `DecisionError`, input guards on confidence/threshold. Model-free and confidence-agnostic so it is hermetically testable - mirrors the C.1 `embed_fn` / C.2 `score_fn` / C.3 scaler seams. |
| Eval harness | `evaluate_librarian.py` | additive `report_decision_accuracy`: fits `T` on positive+hard_negative, runs the live C.1→C.4 decision over the golden set, and reports overall agreement plus **auto-link recall vs review recall** (a single accuracy hides that at `tau=0.80` the softmax top-1 mass of a correct-but-close winner is often ~0.5, so correct positives route to review - the safe direction). Informational, not a hard gate: tuning `tau` is the Week-7 experiment, and flag-based reason codes lag until the SafetyGuard lands. `report_calibration` (W5) untouched. |
| Tests | `decision_engine_test.py` *(new)* | 14 hermetic tests - table-driven over every confidence/flag combination, the inclusive `>=` boundary, all four reason codes, the precedence order, and the input guards. |

## How the pieces connect

```mermaid
flowchart TB
    conf["C.3 calibrated confidence<br/>+ reranked candidates + flags"]
    subgraph C4["C.4 - decision engine (this PR)"]
        rule["decide(): confidence at or above tau ?<br/>AND candidates exist AND no blocking flag"]
        res["DecisionResult<br/>(decision, confidence, cre_ids, reason_code)"]
        rule --> res
    end
    conf --> rule
    res --> link["linked -> LinkProposal (W6b emits)"]
    res --> review["review -> ReviewItem + reason_code<br/>NO_CANDIDATES / ADVERSARIAL_FLAG /<br/>UPDATE_AMBIGUOUS / BELOW_THRESHOLD"]
```

## Results

```
# offline (CI default) - hermetic, no key/DB/model
127 librarian tests passing (113 from W1–W5 + 14 new; 1 skipped)
explicit slice (C.0.5 resolver): 5/5 - gate 100%: PASS

# live (local, against the migrated embedding_vec cache: 428 CRE hub vectors,
# gemini/gemini-embedding-001 dim 3072, ms-marco-MiniLM-L-6-v2, hub-firewall ON)
retrieval recall@20 (C.1): any-hit 285/292 (98%)
rerank top-1     (C.2): 220/292 (75%)
calibration (C.3, 304 rows): T=1.105; ECE 0.046 (calibrated); gate ECE<0.10: PASS
decision (C.4, 319 rows @ tau=0.80): overall 181/319 (57%)
  auto-link recall (expected-linked): 176/314 (56%)
  review recall   (expected-review): 5/5 (100%); reason_code 4/5 (80%)
```

Reading the C.4 line by direction, because a single accuracy hides the story:

- **Review recall 5/5 (100%)** - every chunk that *should* go to a human does. The engine **never wrongly auto-links** something that needs review. For an accuracy *gate*, this is the number that matters most.
- **Auto-link recall 56%** - at `tau=0.80` many correct-but-close positives fall below the bar and route to review (the **safe** direction). This is a conservative starting point; **Week 7's threshold sweep** is exactly the lever that lifts it.
- **reason_code 4/5** - the one miss is a flag-based code that needs the SafetyGuard (not wired; declared).

## What is intentionally not here

- **Envelope emitters + pipeline glue (W6b)** - turning a `DecisionResult` into the RFC `LinkProposal` / `ReviewItem` and wiring C.0→C.4. Ships stacked as `week_6b`.
- **SafetyGuard flags** (`ood` / `conformal` / update-detection) that would populate `adversarial` / `update_ambiguous`.
- **Persisting the fitted `T`** for the live decision path, and **live B→C integration + graph writes (W8)** - the pipeline stays dry-run.

## How to verify locally

```bash
# the new decision-engine tests (hermetic - no key, DB, or model)
python3 -m unittest application.tests.librarian.decision_engine_test
# or the whole librarian suite
python3 -m unittest discover -s application/tests/librarian -p '*_test.py' -t .

# live decision gate (needs a populated cache DB + an embedding-capable LLM).
# after the pgvector migration (#979), migrate a legacy SQLite cache first:
python3 scripts/rewrite_sqlite_embeddings_to_vec.py --db standards_cache.sqlite
python3 scripts/evaluate_librarian.py \
    --dataset application/tests/librarian/fixtures/golden_dataset.json \
    --use_live_embeddings --cache_file standards_cache.sqlite
```

### CODE REVIEW

File: application/utils/librarian/__init__.py [hunk 1]
```diff
@@ -16,7 +16,9 @@
              explicit-link resolution).
   W3 (C.1):  candidate retriever (in-memory + pgvector) + pipeline switch.
   W4 (C.2):  cross-encoder reranker — re-sorts the C.1 shortlist, fills reranked[].
-Calibration + decision routing (C.3-C.4, W5-W6) onward is not built yet.
+  W5 (C.3):  confidence calibration — temperature scaling maps a rerank logit to
+             an honest probability (fit by NLL on the golden set, gated ECE < 0.10).
+Decision routing (C.4, W6) onward is not built yet.
```
Comment (northdpole) [inline application/utils/librarian/__init__.py]: **Nit — package docstring is stale**

This still says "Decision routing (C.4, W6) onward is not built yet." Week 6 adds `decide()` here. Please update the scope blurb to mention C.4 / W6 (and leave W6b emitter/pipeline / W8 writers as not-yet if you prefer).

**Review by northdpole — COMMENTED** [review]
## Maintainer review — Module C Week 6 (#990)

Decision engine looks solid: precedence, inclusive τ, guards, and table tests are clean. No blocking logic bugs in the unique Week-6 surface.

Depends on **#974** for the stack base (please rebase after that gate fix lands). One docstring nit inline.

Non-blocking note: `report_decision_accuracy` intentionally always returns 0 (informational until SafetyGuard + τ tuning) — fine; just don't confuse that with the ECE gate in #974.

File: application/utils/librarian/__init__.py (hunk 1 above)
Comment (PRAteek-singHWY) [reply application/utils/librarian/__init__.py]: Fixed in `86d3d5b5`. Added the W6 (C.4) scope line for `decide()` and moved the not-yet marker down to the W6b emitter/pipeline glue and the W8 queue/graph writers. Will rebase onto `gsocmodule_C_week_5` once #974 lands.

**Comment by PRAteek-singHWY** [comment]
Pushed on top of the docstring fix.

**Carried the #974 fixes onto this branch.** This branch is cumulative off `main`, so it already contained the W5 calibration code including the old `return 0`. All three PRs now show the same corrected code, and the duplicate commits will drop by patch-id when this rebases onto #974 as you asked.

**`report_decision_accuracy` no longer re-derives or re-fits.** It rebuilt the positive + hard_negative calibration set from its own retrieve and rerank pass and fit its own temperature, duplicating what `report_calibration` had done a few lines earlier. On a live run that meant a third pipeline pass over the calibration slices and two independent fits of the same `T`, with nothing guaranteeing the two agreed. Now:

- `calibration_set()` extracts the (shortlist, label) derivation, so the C.3 gate and the C.4 report read the same pairs off the same shared audits.
- `report_calibration` returns `(status, scaler)`; `report_decision_accuracy` takes the fitted scaler, so C.4 thresholds on exactly the `T` the ECE gate measured.
- A degenerate set yields `(1, None)` and C.4 is skipped with a message: no fitted `T` means no honest confidence to threshold, and the run has already failed.
- The live audit set now covers expected-decision rows too. C.4 grades those and they are not confined to the positive/hard_negative slices, so keying them off the calibration slices alone would have dropped them.

**Adds `evaluate_harness_test.py`.** The live reports only run under `--use_live_embeddings`, so nothing exercised their wiring, which is exactly the code that has to share one pipeline pass and one `T` across three reports. A counting stub asserts the pipeline is called once per row and not once per report, that only the two calibration slices enter the fit, and that a degenerate set returns status 1 with no scaler rather than reporting success. Verified that last one fails if the gate is flipped back to 0.

134 librarian tests pass; the hermetic harness run still exits 0.

File: application/tests/librarian/evaluate_harness_test.py [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+
+        audits = harness.live_audits(rows, pipe, pipe)
+
+        self.assertEqual(pipe.retrieve_calls, 2)
+        self.assertEqual(pipe.rerank_calls, 2)
+        self.assertEqual(set(audits), {"p1", "n1"})
+
+        # Three reports read the same audits; none of them may re-run the pipeline.
+        harness.report_retrieval_recall(rows, audits, 10, 5)
+        status, scaler = harness.report_calibration(rows, audits)
```
[the remark this answers is not in the document]
Comment (PRAteek-singHWY) [reply application/tests/librarian/evaluate_harness_test.py]: Fixed in `90992f5f`. Bound to `_status`: this test is about the pipeline not being re-run and the scaler coming back, and the gate outcome on stub logits is not a meaningful assertion. `ruff check` is clean on the file.

File: scripts/evaluate_librarian.py [hunk 3]
```diff
@@ -94,28 +97,21 @@ def predict(section: Section, registry: Set[str], hub: List[HubRep]) -> List[str
...
-      the shortlist, is the #1 candidate an expected CRE? This is the first
-      end-to-end accuracy number for the search path (W4 target >= 0.80).
+):
+    """Construct the live C.1 retriever + C.2 reranker against the OpenCRE DB.
+
+    Live deps are imported lazily so the offline harness needs neither a DB, an
+    embedding model, nor the cross-encoder stack. Called once per run from
+    ``main`` and shared by every live report (recall/top-1 and calibration), so
+    the heavy hub + model load happens a single time and the id-space translation
+    stays in one place.
```
[the remark this answers is not in the document]
Comment (PRAteek-singHWY) [reply scripts/evaluate_librarian.py]: Fixed in `90992f5f`, and it was worse than the three places you named. The module docstring still claimed the semantic path was stubbed. Updated all four: the module docstring, `_build_live_pipeline`'s consumer list, the `--use_live_embeddings` help, and the offline message. They now state that C.3 is the one live report that sets the exit status, since a failed or skipped gate returns nonzero, while C.4 is informational until SafetyGuard and tau tuning land.

**Review by northdpole — CHANGES_REQUESTED** [review]
## Maintainer review — Module C Week 6 (#990) — refresh

**Verdict:** Logic/tests look good; **not mergeable yet** until black is green and the Week-5 base is settled.

### What looks solid
- `decide()` is the right size: pure, model-free, inclusive `>= τ`, top-1 only, frozen `DecisionResult`
- Reason-code precedence matches the RFC story: `NO_CANDIDATES > ADVERSARIAL_FLAG > UPDATE_AMBIGUOUS > BELOW_THRESHOLD`
- Hermetic table tests cover boundary, flags, guards, and precedence
- Declared-degraded SafetyGuard flags (accepted, not produced) is clearly called out — fine for W6
- Decision-accuracy report staying informational (always status 0) is the right call until W7 τ sweep / SafetyGuard

### Blockers
1. **Lint / black** — CI fails on `application/tests/librarian/evaluate_harness_test.py` (`would reformat`). Please run black and push.
2. **Stack** — still depends on **#974** (CHANGES_REQUESTED). Rebase onto `main` after that lands so the Week-6-only surface is what we merge.

### Non-blocking
- Confirm harness tests assert the graded decision outcomes they name (CodeRabbit called this out earlier) — nice-to-have, not a gate for W6.

Once black is green and #974 is in (or this is rebased cleanly), this is approve-ready.

**Comment by northdpole** [comment]
#974 is merged. Please rebase this PR onto `main`, run black (CI was failing on `evaluate_harness_test.py`), and push — happy to re-approve once green.

**Comment by northdpole** [comment]
**Rebase needed (blocking):** this PR is **CONFLICTING** with `main` after #974 merged (~9 commits behind).

Please:
1. Rebase onto latest `main`
2. Run `black` on `application/tests/librarian/evaluate_harness_test.py` (lint still red)
3. Push and ping for re-review

Stack order remains #990 → #991.

**Review by northdpole — APPROVED** [review]
## Re-review (post-rebase)

Prior blocker cleared: **MERGEABLE** on `main`, black/lint clean, all 6 checks green. Thanks for the Slack ping and the stack cleanup.

### Verified
- `decide()` is a small total function with clear reason-code precedence; hermetic table tests look solid.
- Harness shares one retrieve+rerank pass and one fitted `T` across C.3/C.4 reports; C.4 stays informational; degenerate calibration still fails the live run (no greenwash).
- CodeRabbit harness-load note addressed in `dataset_test._load_harness`: `ImportError` with path, **no skip** — correct call.
- Package / calibration docstrings refreshed for W6 scope.

### Optional follow-up (non-blocking)
`evaluate_harness_test.py` still uses a bare `assert _spec and _spec.loader` at import time; aligning it with the same `ImportError(f"... {path}")` pattern as `dataset_test` would be nice consistency, not a merge blocker.

**Approve.** Merge with rebase after CI; then land #991 (stacked on this).
